lib/src/main/java/org/esbtools/eventhandler/PromiseOfPromise.java

```diff
@@ -0,0 +1,69 @@
+/*
+ *  Copyright 2016 esbtools Contributors and/or its affiliates.
+ *
+ *  This file is part of esbtools.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.esbtools.eventhandler;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class PromiseOfPromise<U> implements Promise<U> {
```

Comment:
> Makes having a thenPromise (or equivalent) unnecessary

This was part of the goal.

> Makes it more difficult (if not impossible) to lazily execute computation while in the Future domain

Don't follow. Can you elaborate?

> Is less obvious how to use: I already have a promise from request... why do I need to forward to another?

That's not how I read it. I read it as "the callback for the request is responsible for resolving the promise returned by `process()`", which is exactly the way I think of it... To me, the use of `thenPromise` is less obvious.

> I still don't see a compelling reason why an additional thenPromise API is so egregious.

I simply don't like it because it adds another method which doesn't exist in other promises implementations (AFAIK).

> Spotify's futures-extra library recognized this same restriction, and so they have asyncTransform and syncTransform, analogous to thenPromise and then respectively. Again, we'd probably do well to adopt that family of terminology, but the concepts are basically the same.

This sounds better. I think if we're going in this direction, then we should avoid the term "promise" though, since it departs from the methods available on promises in JavaScript.